### PR TITLE
target/ppc: PowerISA 3.0 VSX instructions

### PR DESCRIPTION
**Pull request by mferst**

### CODE REVIEW

File: target/ppc/fpu_helper.c [hunk 1]
```diff
@@ -2127,14 +2129,49 @@ void helper_##op(CPUPPCState *env, ppc_vsr_t *xt,                             \
...
+}
+
+VSX_MADDQ(XSMADDQP, MADD_FLGS, 0)
+VSX_MADDQ(XSMADDQPO, MADD_FLGS, 1)
+VSX_MADDQ(XSMSUBQP, MSUB_FLGS, 0)
+VSX_MADDQ(XSMSUBQPO, MSUB_FLGS, 1)
+VSX_MADDQ(XSNMADDQP, NMADD_FLGS, 0)
+VSX_MADDQ(XSNMADDQPO, NMADD_FLGS, 1)
+VSX_MADDQ(XSNMSUBQP, NMSUB_FLGS, 0)
+VSX_MADDQ(XSNMSUBQPO, NMSUB_FLGS, 0)
```
Comment (vcoracolombo) [inline target/ppc/fpu_helper.c:2165]: it should be RO=1 here, right?

File: target/ppc/insn32.decode [hunk 2]
```diff
@@ -419,6 +425,32 @@ STXVX           011111 ..... ..... ..... 0110001100 .   @X_TSX
...
+XSNMADDMSP      111100 ..... ..... ..... 10001001 . . . @XX3
+XSNMADDADP      111100 ..... ..... ..... 10100001 . . . @XX3
+XSNMADDMDP      111100 ..... ..... ..... 10101001 . . . @XX3
+XSNMADDQP       111111 ..... ..... ..... 0111000100 .   @X_rc
+
+XSNMSUBASP      111100 ..... ..... ..... 10010001 . . . @XX3
+XSNMSUBMSP      111100 ..... ..... ..... 10011001 . . . @XX3
+XSNMSUBADP      111100 ..... ..... ..... 10110001 . . . @XX3
+XSNMSUBMDP      111100 ..... ..... ..... 10111001 . . . @XX3
+XSNMSUBQP       111111 ..... ..... .... .0111100100 .   @X_rc
```
Comment (vcoracolombo) [inline target/ppc/insn32.decode:452]:
```suggestion
XSNMSUBQP       111111 ..... ..... ..... 0111100100 .   @X_rc
```

File: target/ppc/fpu_helper.c [hunk 3]
```diff
@@ -2127,14 +2129,49 @@ void helper_##op(CPUPPCState *env, ppc_vsr_t *xt,                             \
...
 
-VSX_MADD(xsmadddp, 1, float64, VsrD(0), MADD_FLGS, 1, 0)
-VSX_MADD(xsmsubdp, 1, float64, VsrD(0), MSUB_FLGS, 1, 0)
-VSX_MADD(xsnmadddp, 1, float64, VsrD(0), NMADD_FLGS, 1, 0)
-VSX_MADD(xsnmsubdp, 1, float64, VsrD(0), NMSUB_FLGS, 1, 0)
-VSX_MADD(xsmaddsp, 1, float64, VsrD(0), MADD_FLGS, 1, 1)
-VSX_MADD(xsmsubsp, 1, float64, VsrD(0), MSUB_FLGS, 1, 1)
-VSX_MADD(xsnmaddsp, 1, float64, VsrD(0), NMADD_FLGS, 1, 1)
-VSX_MADD(xsnmsubsp, 1, float64, VsrD(0), NMSUB_FLGS, 1, 1)
+#define VSX_MADDQ(op, maddflgs, ro)                                            \
```
[the remark this answers is not in the document]
Comment (mferst) [reply target/ppc/fpu_helper.c:2158]: Sure. Fixed in bb057fe6010b .

File: target/ppc/fpu_helper.c [hunk 4]
```diff
@@ -2127,14 +2129,49 @@ void helper_##op(CPUPPCState *env, ppc_vsr_t *xt,                             \
...
+    set_float_exception_flags(0, &tstat);                                      \
+    if (ro) {                                                                  \
+        tstat.float_rounding_mode = float_round_to_odd;                        \
+    }                                                                          \
+    t.f128 = float128_muladd(s1->f128, s3->f128, s2->f128, maddflgs, &tstat);  \
+    env->fp_status.float_exception_flags |= tstat.float_exception_flags;       \
+                                                                               \
+    if (unlikely(tstat.float_exception_flags & float_flag_invalid)) {          \
+        float128_maddsub_update_excp(env, s1->f128, s3->f128, s2->f128,        \
+                                 maddflgs, GETPC());                           \
```
[the remark this answers is not in the document]
Comment (mferst) [reply target/ppc/fpu_helper.c]: Yup, forgot to fix the alignment when I changed `tp##` to `float128`. Fixed in bb057fe6010b .

File: target/ppc/fpu_helper.c (hunk 1 above)
Comment (mferst) [reply target/ppc/fpu_helper.c]: Yes, nice catch. Fixed in bb057fe6010b .

File: target/ppc/fpu_helper.c [hunk 5]
```diff
@@ -2127,14 +2129,49 @@ void helper_##op(CPUPPCState *env, ppc_vsr_t *xt,                             \
...
+VSX_MADDQ(XSNMSUBQPO, NMSUB_FLGS, 0)
+
+VSX_MADD(XSMADDDP, 1, float64, VsrD(0), MADD_FLGS, 1, 0)
+VSX_MADD(XSMSUBDP, 1, float64, VsrD(0), MSUB_FLGS, 1, 0)
+VSX_MADD(XSNMADDDP, 1, float64, VsrD(0), NMADD_FLGS, 1, 0)
+VSX_MADD(XSNMSUBDP, 1, float64, VsrD(0), NMSUB_FLGS, 1, 0)
+VSX_MADD(XSMADDSP, 1, float64, VsrD(0), MADD_FLGS, 1, 1)
+VSX_MADD(XSMSUBSP, 1, float64, VsrD(0), MSUB_FLGS, 1, 1)
+VSX_MADD(XSNMADDSP, 1, float64, VsrD(0), NMADD_FLGS, 1, 1)
+VSX_MADD(XSNMSUBSP, 1, float64, VsrD(0), NMSUB_FLGS, 1, 1)
```
[the remark this answers is not in the document]
Comment (mferst) [reply target/ppc/fpu_helper.c]: Fixed in bb057fe6010b .

File: target/ppc/insn32.decode (hunk 2 above)
Comment (mferst) [reply target/ppc/insn32.decode]: Oops. Fixed in bb057fe6010b .

**Review by vcoracolombo — APPROVED** [review]

**Comment by mferst** [comment]
Merged in upstream: https://lists.gnu.org/archive/html/qemu-ppc/2022-03/msg00002.html

**Pull request closed without merging**